design-documents/frontend/server-side-rendering.md

```diff
@@ -0,0 +1,129 @@
+# Server-Side Rendering (SSR)
+
+_This document refers to SSR in the context of a React web application (e.g., PWA). The concept of SSR does not apply to native apps._
+
+_SSR belongs to the BFF layer of [service isolation](https://github.com/magento/architecture/tree/master/design-documents/service-isolation)._
+
+SSR can be a challenge to implement correctly, but the benefits have the potential to provide the unparalleled performance you seek to decrease bounce rate and drive more business.
+
+## Pros
+
+- Faster, first meaningful paint.
+- No need to render initial content via JavaScript on the client.
+- SEO/bot support.
+- Better SEO performance.
+- JavaScript code can be shared universally (full-stack).
+
+## Cons
+
+- User may see some content before it's interactive.
+- Difficult or tedious to implement.
+- May require additional configuration and refactoring.
+- Some features [not yet supported](#not-supported-yet).
+
+## How it works
+
+The basic idea is that we don't want the web browser (client) to be responsible for rendering content via JavaScript, because JavaScript is notoriously slow, especially at rendering DOM and especially on older or mobile devices. Unnecessary rendering also consumes more energy, which has the potential to suck a mobile device's battery dry. :battery:
+
+If we leverage the HTTP server to do as much heavy lifting as we can, some of this heavy lifting (non-sensitive pages) can be cached as well, resulting in a faster, better user experience.
+
+### Without SSR
+
+1. A web browser (client) requests a URL for a particular route (e.g., `/`).
+1. The HTTP server runs the appropriate code for that route.
+   - HTML is partially generated with empty containers that need to be filled and rendered by JavaScript after the initial page load (bad for SEO).
+1. The web browser (client) receives and processes the HTML, but the user may see some empty content that hasn't been filled in yet.
+1. JavaScript is loaded and processed, including the React library.
+1. Ideally, a single GraphQL API request is made to collect the data required to fill in the empty containers with content.
+1. React renders the final HTML into the above-mentioned empty containers by combining the data received with the HTML defined in the React components.
+1. Once content is seen, it becomes immediately interactive.
+
+### With SSR
```

Comment:
We spoke a bit about this offline, but this section is missing some steps that add execution time.

There is a note that says:

> Ideally, a single GraphQL API request is made from this HTTP server, collecting all the data that's required to render the page.
>   - HTML is fully generated [via React methods](https://reactjs.org/docs/react-dom-server.html) with no empty containers. This HTML can be [streamed](https://reactjs.org/docs/react-dom-server.html#rendertonodestream) and cached

How do you know what data to fetch here yet? Unfortunately, for most apps (especially with colocated data fetching) you don't. That's why folks doing SSR today (in the large majority of cases) have to do a 2-pass render. It looks something like this

1. Mount the the initial React App with the proper routing data, based on the request
2. Walk the rendered React tree manually, to collect data requirements for each component.
3. Make an API request on the server side for that data
4. Do a second mount of the React app, this time feeding it both routing data and the API results

For some context, `next.js` uses a custom lifecycle method [`getInitialProps`](https://nextjs.org/learn/basics/fetching-data-for-pages) to handle the data collection from the first pass render.

Reply:
The alternative is not doing the steps mentioned above, and letting all your components render their default states without data.

But that itself has its own costs. Immediately after the first render on the client, you have to go back to the network, get your data, and then re-render the tree from the top down, which could lead to a very different UI.

Reply:
I'm not concered about "for most apps" because this document is about an ideal app. There are various ways to do it, as you pointed out, but that's an implementation detail. We did something similar to Next.js at Nordstrom. The point here is that it can be done. Even if it requires 2 passes, that response is still cacheable.

That said, I'm much more interested in the [GraphQL Code Generator](https://graphql-code-generator.com) approach where you can pluck `gql` tags from the components to form a single query, [like I'm doing here](https://github.com/jedmao/tsx/blob/redux/codegen.yml). This whole topic warrants its own proposal though and I didn't want to derail the gist of this document with those implementation details.

I'm going to address the "critical content" part of your concerns though.